File: src/main/scala/demos/fp/order/smartconstructor/OrderLine.scala
package demos.fp.order.smartconstructor

class OrderLine private (val productId: String, val amount: Int):
    def add(extraAmount: Int): OrderLine =
        val newAmount = this.amount + extraAmount
        if extraAmount <= 0 || this.amount + extraAmount > 10 then throw Exception("Cannot increase order line amount")
        else OrderLine(productId = this.productId, amount = newAmount)

object OrderLine:
    def apply(productId: String, amount: Int): OrderLine =
        if amount <= 0 || amount > 10 then throw Exception("Cannot create order line")
        else new OrderLine(productId = productId, amount = amount)

File: src/main/java/demos/oop/order/smartconstructor/OrderLine.java
package demos.oop.order.smartconstructor;

class OrderLine {
    private String productId;
    private int amount;

    public OrderLine(String productId, int amount) {
        this.productId = productId;
        this.amount = amount;
        this.validate();
    }

    private void validate() {
        if (this.amount <= 0 || this.amount > 10) {
            throw new RuntimeException("Order line amount must be positive and at most 10");
        }
    }

    public void add(int extraAmount) {
        int newAmount = this.amount + extraAmount;
        if (extraAmount <= 0) {
            throw new RuntimeException("Amount to increase must be positive");
        } else {
            this.amount = newAmount;
            this.validate();
        }
    }

    public String getProductId() {
        return productId;
    }

    public int getAmount() {
        return amount;
    }
}
